### DISCUSSION TITLE
Cannot pass URL Encoded string containing '/' as path parameter to GET endpoint

### DISCUSSION THREAD

**Discussion by marshall62** [fastapi/fastapi#8733] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.get('/images/thumbnail/{filename}',
  response_description="Returns a thumbnail image",
  response_class="StreamingResponse",
  responses= {200: {"description": "an image", "content": {"image/jpeg": {}}}})
async def thumbnail_image (filename: str):
```

### Description

I have an endpoint that needs to take a filename (or path to filename) parameter string that could be something like
"myimage.jpg" (which works) or something like "path/to/myimage.jpg" (which doesn't work)  

http://localhost:8000/images/thumbnail/landscapes%2FIMG_1462.jpeg does not enter the endpoint above and gets back a 404

I've tried removing the {filename} path param to form a regular query parameter but it doesn't help.  I am using the Javascript encodeURIComponent to create the string that is being sent as the parameter.

### Operating System

macOS

### FastAPI Version

0.70.1

### Python Version

3.8

**Comment by ypsah** [comment]
From `fastapi`'s logs, I expect something somewhere decodes the URL a bit too early:

```bash
curl -X 'GET' \
  'http://127.0.0.1:8000/images/thumbnail/abc%2Fdef%2Fghi' \
  -H 'accept: application/json'
```

```
INFO:     127.0.0.1:46024 - "GET /images/thumbnail/abc/def/ghi HTTP/1.1" 404 Not Found
```

According to [RFC 3986 2.4](https://datatracker.ietf.org/doc/html/rfc3986#section-2.4):

> When a URI is dereferenced, the components and subcomponents
   significant to the scheme-specific dereferencing process (if any)
   must be parsed and separated before the percent-encoded octets within
   those components can be safely decoded, as otherwise the data may be
   mistaken for component delimiters.

Using a random REST API, I could check that other web frameworks at least don't seem to treat percent-encoded `/` the same way as fastapi when they are part of the `path` component of the URL:

```console
$ curl https://jsonplaceholder.typicode.com/todos/1
{
  "userId": 1,
  "id": 1,
  "title": "delectus aut autem",
  "completed": false
}
$ curl https://jsonplaceholder.typicode.com/todos%2F1
{}
```

For the record, I used a slightly simpler snippet to reproduce:

```python
from fastapi import FastAPI

app = FastAPI()

@app.get('/images/thumbnail/{filename}')
async def thumbnail_image (filename: str) -> str:
    """
    A thumbnail image
    """
    return filename
```

And that one can be made to work by making `filename` a proper query parameter:

```python
from fastapi import FastAPI

app = FastAPI()

@app.get('/images/thumbnails')
async def thumbnail_image (filename: str) -> str:
    """
    A thumbnail image
    """
    return filename
```

Example:

```console
$ curl -X 'GET' \
  'http://127.0.0.1:8000/images/thumbnails?filename=abc%2Fdef%2Fghi' \
  -H 'accept: application/json'
"abc/def/ghi"
```

**Comment by ypsah** [comment]
I'm not 100% sure, but I think the error comes from `uvicorn`, [here](https://github.com/encode/uvicorn/blob/df8b49f6779b9de508416ad3823bac868bb8efcf/uvicorn/protocols/http/h11_impl.py#L170).

`starlette`'s `TestClient` also looks affected [here](https://github.com/encode/starlette/blob/165592fb89ad26383b4a8122b4829bdfe4c64d60/starlette/testclient.py#L164) and [there](https://github.com/encode/starlette/blob/165592fb89ad26383b4a8122b4829bdfe4c64d60/starlette/testclient.py#L180).

In all three places the code looks something like:
```python
self.scope = {
    ...: ...,
    "path": unquote(path),
}
```

Which decodes characters too aggressively IMO.

**Comment by ypsah** [comment]
Looks like the issue was discussed on `uvicorn`: https://github.com/encode/uvicorn/pull/354 and resulted in the following issue being opened on `asgiref` at the time: https://github.com/django/asgiref/issues/87.

Both `asgiref` and `uvicorn` seem to have addressed the issue with the addition of `raw_path`, which makes me think `starlette` might be using `path` in places where it should consider `raw_path` instead.

Won't investigate this any further, but probably worth opening an issue on [`starlette`](https://github.com/encode/starlette) if this is confirmed.

**Comment by ypsah** [comment, marked as answer]
Reading about it again, it looks like `starlette` supports a [`path` convertor](https://www.starlette.io/routing/#path-parameters).

The following should do what you want:

```python
from fastapi import FastAPI

app = FastAPI()

@app.get('/images/thumbnail/{filename:path})
async def thumbnail_image(filename: str) -> str:
    """
    A thumbnail image
    """
    return filename
```

**Comment by Yuutakasan** [comment]
We are developing a bulletin board system that uses APIs, and when the path parameter contains a URL (which of course contains /) When the path parameter contains a URL (which of course contains a /), the / is decoded into a / before the routing interprets it, resulting in a 404 because the path does not match. The problem occurs even if you urlencode and convert the / to %2F beforehand.

**Comment by d13g0** [comment]
Hi, my workaround to this issue was passing a base64 encoded variable in the querystring. My FastAPI endpoint looks like this:

```python
from base64 import urlsafe_b64decode

@router.get('/my_route')
def my_endpoint(url:str = None):
    
    if url is None:
           raise HTTPException(...)
           
    parsed_url = urlsafe_b64decode(url)
    crud.operation(parsed_url,...)
```

**Comment by ChiliJohnson** [comment]
I opened a PR into Starlette which makes its router use ASGI's new `raw_path` instead of the pre-decoded path, so that it can differentiate between `%2F` and `/`, especially when using the path converter: encode/starlette#1828 and encode/starlette#1827

There doesn't seem to be much discussion happening over there, but you can try checking out that PR branch and see if it helps! If it does help you, mentioning your use case in that thread might be a good bump for the discussion around getting it merged
